### ISSUE TITLE
Deleting column with index: InvalidArgumentException: The specified index name '....'

### ISSUE THREAD

**Issue by filisko** [odan/phinx-migrations-generator#65]
Hi,

1. Generated for first time migration file with schema file from **development** DB (90 tables).
2. Run the migration on **testing** DB.
3. Removed a column with foreign key on **development** DB and generated migration file.
4. Run the migration on **testing** DB, and I got the error.

Could that be because bad conventions are being used or something like that?

I also realized that db transactions are not being used, right? Wouldn't that be a good idea?
phinxlog table does not get updated with the failing migration, but some fields actually get changed

```
20190926071410 DbChange4138451195d8c654158c26: migrating
PDOException: SQLSTATE[HY000]: General error: 1553 Cannot drop index 'fk_activity_payments1': needed in a foreign key constraint in /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Adapter/PdoAdapter.php:168
Stack trace:
#0 /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Adapter/PdoAdapter.php(168): PDO->exec('ALTER TABLE `ac...')
#1 /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Util/AlterInstructions.php(126): Phinx\Db\Adapter\PdoAdapter->execute('ALTER TABLE `ac...')
#2 /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Adapter/PdoAdapter.php(573): Phinx\Db\Util\AlterInstructions->execute('ALTER TABLE `ac...', Array)
#3 /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Adapter/PdoAdapter.php(932): Phinx\Db\Adapter\PdoAdapter->executeAlterSteps('activity', Object(Phinx\Db\Util\AlterInstructions))
#4 /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Adapter/AdapterWrapper.php(497): Phinx\Db\Adapter\PdoAdapter->executeActions(Object(Phinx\Db\Table\Table), Array)
#5 /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Adapter/TimedOutputAdapter.php(406): Phinx\Db\Adapter\AdapterWrapper->executeActions(Object(Phinx\Db\Table\Table), Array)
#6 /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Plan/Plan.php(146): Phinx\Db\Adapter\TimedOutputAdapter->executeActions(Object(Phinx\Db\Table\Table), Array)
#7 /phinx/vendor/cakephp/collection/CollectionTrait.php(64): Phinx\Db\Plan\Plan->Phinx\Db\Plan\{closure}(Object(Phinx\Db\Plan\AlterTable), 0)
#8 /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Plan/Plan.php(147): Cake\Collection\Collection->each(Object(Closure))
#9 /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Table.php(721): Phinx\Db\Plan\Plan->execute(Object(Phinx\Db\Adapter\TimedOutputAdapter))
#10 /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Table.php(628): Phinx\Db\Table->executeActions(true)
#11 /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Table.php(687): Phinx\Db\Table->update()
#12 /phinx/migrations/20190926071410_db_change4138451195d8c654158c26.php(53): Phinx\Db\Table->save()
#13 /phinx/vendor/robmorgan/phinx/src/Phinx/Migration/Manager/Environment.php(125): DbChange4138451195d8c654158c26->change()
#14 /phinx/vendor/robmorgan/phinx/src/Phinx/Migration/Manager.php(391): Phinx\Migration\Manager\Environment->executeMigration(Object(DbChange4138451195d8c654158c26), 'up', false)
#15 /phinx/vendor/robmorgan/phinx/src/Phinx/Migration/Manager.php(365): Phinx\Migration\Manager->executeMigration('prod', Object(DbChange4138451195d8c654158c26), 'up', false)
#16 /phinx/vendor/robmorgan/phinx/src/Phinx/Console/Command/Migrate.php(123): Phinx\Migration\Manager->migrate('prod', 20190926071410, false)
#17 /phinx/vendor/symfony/console/Command/Command.php(255): Phinx\Console\Command\Migrate->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#18 /phinx/vendor/symfony/console/Application.php(915): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#19 /phinx/vendor/symfony/console/Application.php(272): Symfony\Component\Console\Application->doRunCommand(Object(Phinx\Console\Command\Migrate), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#20 /phinx/vendor/robmorgan/phinx/src/Phinx/Console/PhinxApplication.php(96): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#21 /phinx/vendor/symfony/console/Application.php(148): Phinx\Console\PhinxApplication->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#22 /phinx/vendor/odan/phinx-migrations-generator/bin/phinx-migrations(20): Symfony\Component\Console\Application->run()
#23 {main}
```

```
== 20190926065526 DbChange7088656665d8c60dcc2c70: migrating
InvalidArgumentException: The specified index name 'fk_users_id_foreign3_idx' does not exist in /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Adapter/MysqlAdapter.php:654
Stack trace:
#0 /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Adapter/PdoAdapter.php(872): Phinx\Db\Adapter\MysqlAdapter->getDropIndexByNameInstructions('activity', 'fk_users_id_for...')
#1 /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Adapter/AdapterWrapper.php(497): Phinx\Db\Adapter\PdoAdapter->executeActions(Object(Phinx\Db\Table\Table), Array)
#2 /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Adapter/TimedOutputAdapter.php(406): Phinx\Db\Adapter\AdapterWrapper->executeActions(Object(Phinx\Db\Table\Table), Array)
#3 /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Plan/Plan.php(146): Phinx\Db\Adapter\TimedOutputAdapter->executeActions(Object(Phinx\Db\Table\Table), Array)
#4 /phinx/vendor/cakephp/collection/CollectionTrait.php(64): Phinx\Db\Plan\Plan->Phinx\Db\Plan\{closure}(Object(Phinx\Db\Plan\AlterTable), 'activity')
#5 /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Plan/Plan.php(147): Cake\Collection\Collection->each(Object(Closure))
#6 /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Table.php(721): Phinx\Db\Plan\Plan->execute(Object(Phinx\Db\Adapter\TimedOutputAdapter))
#7 /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Table.php(628): Phinx\Db\Table->executeActions(true)
#8 /phinx/vendor/robmorgan/phinx/src/Phinx/Db/Table.php(687): Phinx\Db\Table->update()
#9 /phinx/migrations/20190926065526_db_change7088656665d8c60dcc2c70.php(30): Phinx\Db\Table->save()
#10 /phinx/vendor/robmorgan/phinx/src/Phinx/Migration/Manager/Environment.php(125): DbChange7088656665d8c60dcc2c70->change()
#11 /phinx/vendor/robmorgan/phinx/src/Phinx/Migration/Manager.php(391): Phinx\Migration\Manager\Environment->executeMigration(Object(DbChange7088656665d8c60dcc2c70), 'up', false)
#12 /phinx/vendor/robmorgan/phinx/src/Phinx/Migration/Manager.php(365): Phinx\Migration\Manager->executeMigration('prod', Object(DbChange7088656665d8c60dcc2c70), 'up', false)
#13 /phinx/vendor/robmorgan/phinx/src/Phinx/Console/Command/Migrate.php(123): Phinx\Migration\Manager->migrate('prod', 20190926065526, false)
#14 /phinx/vendor/symfony/console/Command/Command.php(255): Phinx\Console\Command\Migrate->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#15 /phinx/vendor/symfony/console/Application.php(915): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#16 /phinx/vendor/symfony/console/Application.php(272): Symfony\Component\Console\Application->doRunCommand(Object(Phinx\Console\Command\Migrate), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#17 /phinx/vendor/robmorgan/phinx/src/Phinx/Console/PhinxApplication.php(96): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#18 /phinx/vendor/symfony/console/Application.php(148): Phinx\Console\PhinxApplication->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#19 /phinx/vendor/odan/phinx-migrations-generator/bin/phinx-migrations(20): Symfony\Component\Console\Application->run()
#20 {main}
```

**Comment by odan** [comment]
Hi @filisko :-)

> Could that be because bad conventions are being used or something like that?

I guess not.

> Cannot drop index 'fk_activity_payments1': needed in a foreign key constraint 

Every FK needs an index. If we try to delete an index of table A, but a FK from table B has a reference to this column in table A, we get this error. [InnoDB and FOREIGN KEY Constraints](https://dev.mysql.com/doc/refman/5.6/en/innodb-foreign-key-constraints.html)

> I also realized that db transactions are not being used, right? Wouldn't that be a good idea?

Definition language (DDL) statements, such as those that create or drop databases, those that create, drop, or alter tables or stored routines cannot be rolled back. In other words: All table structure changes are not "covered" by transactions. Only data changes can be reverted by transactions.

For this reason, it would not make sense to add transactions for DDL migrations. For all other migrations that contain DML (INSERT, UPDATE, and DELETE statements), a transaction would of course be useful. 

[Statements That Cannot Be Rolled Back](https://dev.mysql.com/doc/refman/5.7/en/cannot-roll-back.html)

Could you please provide the generated diff from the second migration?

**Comment by filisko** [comment]
Sorry, on point 3 I actually delete first (manually) the foreign key (and that deletes the index) and then the column.

**Comment by odan** [comment]
Okay, I'm trying to reproduce the error on my machine.

**Comment by odan** [comment]
This looks like a bug in phinx, because the generated sql by phinx looks not ok.

This code...
```php
   ->removeColumn('payment_id')
   ->removeIndexByName("fk_activity_payments1")
   ->dropForeignKey('fk_activity_payments1')
```

generates this SQL (I changed the names):

```sql
ALTER TABLE `table1` DROP COLUMN `table2_id`
```

Error:

```
SQLSTATE[HY000]: General error: 1553 Cannot drop index 'table1_table2_id': needed in a foreign key constraint
```

The expected SQL would be:

```sql
ALTER TABLE `table1` DROP COLUMN `table2_id`, DROP INDEX `table1_table2_id`, DROP FOREIGN KEY `table1_table2_id`; 
```

This looks like a bug in phinx that has already reported and fixed here [#1514](https://github.com/cakephp/phinx/issues/1514), but the sql is still not correct.

I think we should report this bug in phinx.

**Comment by filisko** [comment]
Oh, really? I was also looking into that code but didn't test the SQL output, nice!

thank you

**Comment by odan** [comment]
The fix has been merged into Phinx 0.11.2. Update to the latest phinx version and try it again.
